java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -131,33 +131,48 @@ public DictionaryProvider takeDictionaryOwnership() {
     return provider;
   }
 
+  /**
+   * Get the descriptor for this stream. Only applicable on the server side of a DoPut operation. Will block until the
+   * client sends the descriptor.
+   */
   public FlightDescriptor getDescriptor() {
+    // This blocks until the schema message (with the descriptor) is sent.
+    getRoot();
     return descriptor;
   }
 
   /**
    * Closes the stream (freeing any existing resources).
    *
-   * <p>If the stream is isn't complete and is cancellable this method will cancel the stream first.</p>
+   * <p>If the stream isn't complete and is cancellable, this method will cancel the stream first.</p>
    */
   public void close() throws Exception {
-    if (!completed && cancellable != null) {
-      cancel("Stream closed before end.", null);
+    RuntimeException cancelException = null;
+    try {
+      // cancel() can throw, but we still want to clean up resources.
+      if (!completed && cancellable != null) {
+        cancel("Stream closed before end.", /* no exception to report */ null);
+      }
+    } catch (RuntimeException e) {
```

Comment:
just make this a closeable as well, will be even simpler since you won't have to deal with the suppressed pattern below (also, not sure the cancel exception should be the suppressed, maybe the initial exception) :)

Reply:
Good idea, thank you!